Selection widget

fiskus opened the pull request:
# Description


# TODO

<!-- Remove items that are irrelevant to this PR -->

- [ ] Unit tests
- [ ] Automated tests (e.g. Preflight)
- [ ] Confirm that this change meets security best practices and does not violate the security model
- [ ] Documentation
    - [ ] [Python: Run `build.py`](../gendocs/build.py) for new docstrings
    - [ ] JavaScript: basic explanation and screenshot of new features
    - [ ] Markdown somewhere in docs/**/*.md that explains the feature to end users (said .md files should be linked from SUMMARY.md so they appear on https://docs.quiltdata.com)
    - [ ] Markdown docs for developers
- [ ] [Changelog](CHANGELOG.md) entry (skip if change is not significant to end users, e.g. docs only)

fiskus commented on catalog/app/containers/Bucket/Dir.tsx:
```diff
@@ -210,35 +152,24 @@ function DirContents({ response, locked, bucket, path, loadMore }: DirContentsPr
-        onSelectionChange={handleSelectionModelChange}
-        selection={selection}
+        onSelectionChange={locked ? undefined : onSelection}
+        selection={locked ? undefined : selection}
```
Yes, and I wanted to use skeleton to avoid it

nl0 replied:
why not just keep what's on the screen? skeleton will probably introduce another jump (unless you keep the element height)

fiskus replied:
Thanks, I reiterated working on this issue and found solution. I thought some state is unmanageable here, but it's not

fiskus commented on catalog/app/containers/Bucket/PackageDialog/S3FilePicker.tsx:
```diff
@@ -264,8 +304,8 @@ export function Dialog({
               loadMore={loadMore}
-              selection={selection}
-              onSelectionChange={setSelection}
+              selection={Selection.getDirectorySelection(selection, bucket, path)}
```
Yes, it's basically `selection[`bucket+path`]`, but isolated in a function

fiskus commented:
> is the selection "consumed" after the package is created or does it just stay there?
* Selection is consumed in S3FilePicker, because users select files in popup, and every time they "Add files from bucket" they start from zero selection
* Selection stays there if users stay on the bucket page. If they select something, create package and click "Close" on Success Dialog (not "Browse" new package), then they probably want to continue to work with files and selection.

fiskus commented:
Done/Fixed:

* "Continue" is renamed to "Close"
* Reduced empty space around popup buttons (aka selection toolbar)
* Handle long names and long URLs
* Preserve checkboxes, so UI doesn't "jump"
* Show confirmation when navigating to a file
* Added ".." in the S3FilePicker similarly to Bucket
* Fixed breadcrumbs paddings
* Fixed close button size
* Handle long names in FilesInput
* Fix resolving paths when "Add files from bucket" 
* Buttons and button with badge: made margins equal


Partially:

* I can't remove checkbox for "..", but I made it looks disabled


Questions:

* Selection consume
  S3FilePicker: It's consumed, because it's in popup
  Bucket: It's preserved for another action
* Clear files
  S3FilePicker: selected files removed, because user adds files
  Bucket: selected files untouched, because user creates package with these initial files

Not Done:

1. Buttons at the top. I tried to move them down, but it's even worse.
2. Switch S3Filepicker's tree and selection. Good idea, but I'm afraid I can't fit into horizon.
3. Selection counter at the bottom.
  In the top is consistently with counter on Bucket tab. And close button at the same position as counter button.

fiskus commented:
Also, things to consider:
* tsified "app/constant/routes"
* mocked "p-limit" for jest, because it's ES module and doesn't work in our environment

fiskus commented on catalog/app/containers/Bucket/Dir.tsx:
```diff
@@ -249,6 +180,69 @@ function DirContents({ response, locked, bucket, path, loadMore }: DirContentsPr
+  const [opened, setOpened] = React.useState(false)
+  const open = React.useCallback(() => setOpened(true), [])
+  const close = React.useCallback(() => setOpened(false), [])
+  React.useEffect(() => close(), [close, location])
```
`location` object doesn't change on re-renders

nl0 reviewed (dismissed):
i haven't tested this live, since it's not deployed anywhere, but he code changes look good

nl0 commented:
> Switch S3Filepicker's tree and selection. Good idea, but I'm afraid I can't fit into horizon.

might be worth it to make it a stretch goal in this case, like, maybe try to implement it during broad work after the initial version is merged

fiskus commented on catalog/app/constants/routes.ts:
```diff
@@ -0,0 +1,306 @@
+
+export type ActivateArgs = [options: { registryUrl: string; token: string }]
+
+export const activate: Route<ActivateArgs> = {
```
We, probably, have to use some helper, when we migrate to `react-router-dom@v6` anyway. I didn't dig into it completely, but the current routes constants don't just work because `v6` handles nested routes differently

nl0 replied:
> handles nested routes differently

it can handle absolute routes just as well, afaict.
the problem is v6 doesn't support regexes and has some issues with trailing slashes, so it won't be a seamless upgrade, and a helper won't fix it -- we would have to adjust the routes themselves

fiskus commented:
TODO/stretch goals:
* Move `NoArgs` to default `Route`'s argument
* Refactor `routes`:
  * consider converting objects to functions with two arguments
  * tsify embed routes
  * export the whole route maps for catalog and for embed
* rearrange selection widget in S3FilePicker: use tab-like experience instead of popup

fiskus commented:
Sorry, I cleared approve by small changes:
* fixed matching directory root route for navigation prompt
* moved NoArgs to default params
* deleted CHANGELOG entry from predecessor's PR

nl0 commented on docs/CHANGELOG.md:
```diff
@@ -29,6 +29,7 @@ Entries inside each section should be ordered by type:
 * [Changed] Enable user selection in perspective grids ([#3453](https://github.com/quiltdata/quilt/pull/3453))
 * [Changed] Hide columns without values in files listings ([#3512](https://github.com/quiltdata/quilt/pull/3512))
 * [Changed] Enable `allow-same-origin` for iframes in browsable buckets ([#3516](https://github.com/quiltdata/quilt/pull/3516))
+* [Changed] Preserve selection in Bucket tab when navigate or filter ([#3527](https://github.com/quiltdata/quilt/pull/3527))
```
this doesn't seem to reflect what's this change is actually about
